# vim: syntax=python tabstop=4 expandtab
# coding: utf-8

__author__ = "Joel Ås, Massimiliano Volpe, Chelsea Ramsin & Lauri Mesilaakso"
__copyright__ = "Copyright 2022, Joel Ås, Massimiliano Volpe, Chelsea Ramsin & Lauri Mesilaakso"
__email__ = "massimiliano.volpe@liu.se, chelsea.ramsin@regionostergotland.se & lauri.mesilaakso@regionostergotland.se"
__license__ = "GPL-3"


include: "rules/common.smk"
include: "rules/reform_genomic_region.smk"
include: "rules/depth_of_coverage.smk"
include: "rules/get_interaction_guidelines.smk"
include: "rules/get_clinical_guidelines.smk"
include: "rules/get_target_variants.smk"
include: "rules/variant_annotator.smk"
include: "rules/variant_filtration.smk"


report: "report/workflow.rst"


rule all:
    input:
        unpack(compile_output_list),


module snv_indels:
    snakefile:
        github("hydra-genetics/snv_indels", path="workflow/Snakefile", tag="v0.2.0")
    config:
        config


use rule bed_split from snv_indels as snv_indels_bed_split with:
    input:
        bed="pgx/reform_genomic_region/get_padded_bed/padded_bait_interval.bed",


use rule haplotypecaller from snv_indels as snv_indels_haplotypecaller


module alignment:
    snakefile:
        github("hydra-genetics/alignment", path="workflow/Snakefile", tag="v0.2.0")
    config:
        config


use rule samtools_extract_reads from alignment as alignment_samtools_extract_reads


use rule picard_mark_duplicates from alignment as alignment_picard_mark_duplicates


use rule samtools_index from alignment as alignment_samtools_index
